language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,370 @@
+        -- Note that in case of NonRec, tycon must not be visible in the constructor's result type.
+        -- However, we cannot split&take the result type *without* first running the normalizer.
+        -- To fix this we use the whole scope and only later do an extra correct scope-check inside checkFixScope
+        withTyVarDecls (tycon:tyargs) .
+            checkStarNorm
+        >=> through checkFixScope
+        >=> through checkConRes
+    ) where
+        checkFixScope nTy = case r of
+            NonRec ->
+                -- now we make sure that dataconstructor is not self-recursive, i.e. funargs don't contain tycon
+                withTyVarDecls tyargs $ -- tycon not in scope here
+                      -- OPTIMIZE: we use inferKind for scope-checking, but a simple ADT-traversal would suffice
+                      for_ (funTyArgs nTy) inferKindM
+            _ -> pure ()
+        checkConRes :: Type TyName uni ann -> TypeCheckM uni e ()
+        checkConRes nTy = do
+            -- We already checked that datacons' type is *-kinded (using checkStarNorm), but this is not enough:
+            -- we must also check that its result type is EXACTLY `[[TypeCon tyarg1] ... tyargn]`
+            when (funResultType nTy /= expectedConRes) .
+                throwing _TypeErrorExt $ MalformedDataConstrResType ann expectedConRes
+        expectedConRes = mkAppliedTyCon dt -- [[TypeCon tyarg1] ... tyargn]
+
+-- | For a single binding, returns all newly-declared variables next to their normalized types
+-- There is one implicitly-declared variable: the destructor.
+-- NB: [Input] Assumes that the input is globally-unique and any input subtypes are prior normalized
+-- NB: [Output] Uses renaming to preserve global uniqueness at the output, so the outputs is ready to be put into an env
+bindingNewVars :: MonadQuote m
+               => Normalized (Binding TyName Name uni ann)
+               -> m [(Name, Normalized (Type TyName uni ()))]
+bindingNewVars = unNormalized >>> \case
+  -- @effectfully: is it needed to rename the termbind's normalized-type? I don't think so.
+  TermBind _ _ (VarDecl _ n ty) _rhs ->
+      pure [(n, Normalized $ void ty)]
+
+  DatatypeBind _ dt@(Datatype ann _ tyargs desName vdecls) -> do
+      -- build the patternfunctor's "normalized" type and from it the "normalized" destructor
+      pf <- mkDatatypePatternFunctor ann dt
+
+      let -- build the normalized constructors
+          constrs = fmap (\(VarDecl _ constrName constrTy) -> (constrName, PIR.mkIterTyForall tyargs constrTy))
+                         vdecls
+          des = (desName, mkDestructorTy ann pf dt)
+          newStructors = des:constrs
+
+      -- rename all *-structors before we put them into the env, to make sure we preserve global uniqueness
+      forMOf (traversed._2) newStructors $
+          Normalized . void <<$>> rename
+
+  _ -> pure [] -- typebinds don't introduce vardecls
```

Comment:
Don't use `_` here and in other places. Match on the remaining cases explicitly, so that it's completely transparent how each possible case is handled.